applications/GeoMechanicsApplication/custom_processes/geo_integration_values_extrapolation_to_nodes_process.cpp

```diff
@@ -0,0 +1,242 @@
+        if (rElem.IsActive()) {
+            auto& r_this_geometry = rElem.GetGeometry();
+            const GeometryData::IntegrationMethod this_integration_method = rElem.GetIntegrationMethod();
+            const Matrix extrapolation_matrix =
+                GetExtrapolationMatrix(rElem, r_this_geometry, this_integration_method);
+
+            const SizeType integration_points_number =
+                r_this_geometry.IntegrationPoints(this_integration_method).size();
+            AddIntegrationContributionsForAllVariableLists(rElem, integration_points_number, extrapolation_matrix);
+        }
+    });
+
+    // AssembleNodalDataForAllVariableLists();
+}
+
+void GeoIntegrationValuesExtrapolationToNodesProcess::AssembleNodalDataForAllVariableLists()
+{
+    AssembleNodalData(mDoubleVariables);
+    AssembleNodalData(mArrayVariables);
+    AssembleNodalData(mVectorVariables);
+    AssembleNodalData(mMatrixVariables);
+}
+
+void GeoIntegrationValuesExtrapolationToNodesProcess::AddIntegrationContributionsForAllVariableLists(
+    Element& rElem, SizeType NumberOfIntegrationPoints, const Matrix& rExtrapolationMatrix)
+{
+    for (const auto p_var : mDoubleVariables) {
+        AddIntegrationContributionsToNodes(rElem, *p_var, rExtrapolationMatrix,
+                                           NumberOfIntegrationPoints, AtomicAdd<double>);
+    }
+    for (const auto p_var : mArrayVariables) {
+        AddIntegrationContributionsToNodes(rElem, *p_var, rExtrapolationMatrix,
+                                           NumberOfIntegrationPoints, AtomicAdd<double, 3>);
+    }
+    for (const auto p_var : mVectorVariables) {
+        AddIntegrationContributionsToNodes(rElem, *p_var, rExtrapolationMatrix,
+                                           NumberOfIntegrationPoints, AtomicAddVector<Vector, Vector>);
+    }
+    for (const auto p_var : mMatrixVariables) {
+        AddIntegrationContributionsToNodes(rElem, *p_var, rExtrapolationMatrix,
+                                           NumberOfIntegrationPoints, AtomicAddMatrix<Matrix, Matrix>);
+    }
+}
+
+Matrix GeoIntegrationValuesExtrapolationToNodesProcess::GetExtrapolationMatrix(
+    const Element& rElement, GeometricalObject::GeometryType& rGeometry, const GeometryData::IntegrationMethod& rIntegrationMethod)
+{
+    if (!ExtrapolationMatrixIsCachedFor(rElement)) {
+        CacheExtrapolationMatrixFor(
+            rElement, mpExtrapolator->CalculateElementExtrapolationMatrix(rGeometry, rIntegrationMethod));
+    }
+
+    return GetCachedExtrapolationMatrixFor(rElement);
+}
+
+bool GeoIntegrationValuesExtrapolationToNodesProcess::ExtrapolationMatrixIsCachedFor(const Element& rElement) const
+{
+    return mExtrapolationMatrixMap.count(typeid(rElement).hash_code()) > 0;
+}
+
+Matrix GeoIntegrationValuesExtrapolationToNodesProcess::GetCachedExtrapolationMatrixFor(const Element& rElement)
+{
+    return mExtrapolationMatrixMap[typeid(rElement).hash_code()];
+}
```

Comment:
I believe we can make this member function `const` if we use the `at` member rather than `operator[]`:
```suggestion
Matrix GeoIntegrationValuesExtrapolationToNodesProcess::GetCachedExtrapolationMatrixFor(const Element& rElement) const
{
    return mExtrapolationMatrixMap.at(typeid(rElement).hash_code());
}
```

Reply:
Done (could be done already due to the `mutable`, but as discussed it's nicer when this one throws